content/posts/2019/2019-07-05-littlest-cuda-hub.md

```diff
@@ -0,0 +1,148 @@
+
+- At step 6 we selected a `p3.8xlarge` instance as it comes with four Nvidia Tesla V100 GPUs. 
+- At step 7 we skipped entering the user data (bootstrapping script) as we wanted to run this ourselves manually later.
+- At step 8 we selected 100GB to give ourselves reasonable space for data.
+- Stop after step 16. As we skipped step 7, steps 17-19 do not apply and we need to do some manual configuration before continuing.
+
+### Drivers
+
+Next you need to ssh to the box with the ssh key you selected during setup. In order to make use of GPU enabled libraries we needed to install the NVidia CUDA drivers.
+
+- Visit https://www.nvidia.co.uk/Download/index.aspx
+  - Select the appropriate GPU for the instance (in this case Tesla V100)
+  - Select Linux 64-bit Ubuntu 18.04 
+    - _WARNING: Do not select the generic linux 64-bit driver as it will not work on ubuntu_
+  - Copy the download link and run `wget` on the EC2 instance to download the `.deb` file.
+  - Install your `.deb` file on the server (in our case we ran `dpkg -i nvidia-diag-driver-local-repo-ubuntu1804-418.67_1.0-1_amd64.deb` but your driver file name may differ)
+- This will have installed a new repository but not the actual drivers so let's also do that.
+  - `apt-get update`
+  - `apt-get install cuda-drivers`
+  - We now need to `reboot` for the drivers to be loaded into the kernel
+
+### JupyterHub
+
+Now that we have drivers in place and have rebooted we can install the Littlest JupyterHub as instructed in the documentation. In step 7 we skipped entering the installation command in the user data field, so let's run that manually now.
+
+```
+$ curl https://raw.githubusercontent.com/jupyterhub/the-littlest-jupyterhub/master/bootstrap/bootstrap.py | sudo python3 - --admin <admin-user-name>
+```
+
+Select an admin username that you will remember as you will need that the first time you log in to JupyterHub.
+
+We should now be able to return to the tljh documentation and continue from step 20 by visiting the server's IP address in the browser.
+
+Once you finish this I also recommend upgrading the version of Jupyter Lab to the latest by running `sudo -E pip install -U jupyterlab` from within Jupyter Lab.
+
+### RAPIDS
+
+Now that we have a working enviroment we can get on with the interesting bit of installing the packages from RAPIDS. There is a [useful utility](https://rapids.ai/start.html) on the RAPIDS website for constructing your installation command.
+
+You should select:
+- Conda
+- RAPIDS Stable
+- All packages
+- Python 3.7
+- CUDA 10.0
+
+Then copy the command, prefix it with `sudo -E`, add `cupy` to the end and run it in Jupyter Lab. For me this command looked like this.
+
+```
+$ sudo -E conda install -c nvidia -c rapidsai -c numba -c conda-forge -c pytorch -c defaults \
+cudf=0.8 cuml=0.8 cugraph=0.8 python=3.7 cudatoolkit=10.0 cupy
```

Comment:
I wonder if we might drop the 0.8 bits here for RAPIDS

I suspect that the code examples in this post will keep for future versions, and this seems like the kind of post that people might refer to a year or two from now.